src/Xrefcheck/Verify.hs

```diff
@@ -269,14 +273,24 @@ verifyReference
                 givenAnchors
           in throwError $ AnchorDoesNotExist anchor similarAnchors
 
-checkExternalResource :: VerifyConfig -> Text -> IO (VerifyResult VerifyError)
-checkExternalResource VerifyConfig{..} link
+checkExternalResource
+  :: VerifyConfig
+  -> IORef (Map Text $ Either VerifyError ())
+  -> Text
+  -> IO (VerifyResult VerifyError)
+checkExternalResource VerifyConfig{..} externalCheckedRef link
   | isIgnored = return mempty
   | doesReferLocalhost = return mempty
   | otherwise = fmap toVerifyRes $ do
-      makeRequest HEAD 0.3 >>= \case
-        Right () -> return $ Right ()
-        Left   _ -> makeRequest GET 0.7
+      externalChecked <- readIORef externalCheckedRef
+      if M.member link externalChecked
+      then return $ externalChecked M.! link
+      else do
+        result <- makeRequest HEAD 0.3 >>= \case
+          Right () -> return $ Right ()
+          Left   _ -> makeRequest GET 0.7
+        writeIORef externalCheckedRef (M.insert link result externalChecked)
+        return result
```

Comment:
This logic does not guarantee that a link won't be checked twice by two threads.

For example, imagine we have two references `r1` and `r2`, pointing to the same link `l`.
Now consider this sequence of events:
1. `verifyRepo` spawns two threads, `t1` and `t2`. `t1` will check `r1`, and `t2` will check `r2`.
2. `t1` uses `readIORef` to read the map, and checks whether the map contains `l`. It doesn't, so it proceeds with the check.
3. `t2` uses `readIORef` to read the map, and checks whether the map contains `l`. It doesn't, so it proceeds with the check.
4. `t1` finishes and writes the result to the `IORef`
5. `t2` finishes and writes the result to the `IORef`

Reply:
There's actually another bug here:

Imagine we have two references `r1` and `r2`, pointing to different links `l1` and `l2`.
Now consider this sequence of events:
1. `verifyRepo` spawns two threads, `t1` and `t2`. `t1` will check `r1`, and `t2` will check `r2`.
2. `t1` uses `readIORef` to read the **empty** map, and checks whether the map contains `l1`. It doesn't, so it proceeds with the check.
3. `t2` uses `readIORef` to read the **empty** map, and checks whether the map contains `l2`. It doesn't, so it proceeds with the check.
4. `t1` finishes, modifieds the **empty** map by adding an entry for `l1`, and writes the result to the `IORef` using `writeIORef`
5. `t2` finishes, modifieds the **empty** map by adding an entry for `l2`, and writes the result to the `IORef` using `writeIORef`

In the end, the IORef might contain a map with only 1 entry (instead of 2).

The problem here is that different thread's `read` and `write` operations might overlap, so they'll overwrite each other's work.

Reply:
My 50 cents :sweat_smile: : I think it would be helpful if you write down a design of the proper solution. It is likely simple but may need some relatively advanced knowledge to come to.

(I can think over it if you agree but are currently busy with other stuff)

Reply:
> are currently busy with other stuff

Yeah same :sweat_smile: 

>  I think it would be helpful if you write down a design of the proper solution. It is likely simple but may need some relatively advanced knowledge to come to.

Agreed! @breakerzirconia please try to come up with a design (maybe looking into IORef's `atomic*` operations and `MVar` _may_ be helpful), we can continue this discussion in Slack, feel free to ask any questions at all. When we agree on a plan, then we can implement it ^^